### PR TITLE
Fix NPE on XData (v5.0.2) for Gemini Streaming Exchange when args is empy

### PR DESCRIPTION
**Pull request by dazito**
Respect the homeCurrency configuration field when using a StreamingExchange class

On the streaming class, if an exchange has the homeCurrency set in the configuration yaml file this was not taking in consideration (missing a call `to exchangeService.convertExchangePair(exchange, pair)`)

For `GeminiStreamingMarketDataService` on v5.0.2 we must pass on a value to the args parameter or it will throw a null pointer exception (I fixed it on the XChange side, please see: https://github.com/knowm/XChange/pull/3761). So as soon as the new version is released we can clean up this if else code.

On line 74 (`exchange.getStreamingMarketDataService().getTicker(exchangeService.convertExchangePair(exchange, pair), 10);`) I choose the number 10 for no special reason. Please let me know if it should be a different number and I will change it.

### CODE REVIEW

**Comment by scionaltera** [comment]
Looks like your PR with XChange got merged into `develop` so that NPE will be fixed in their next release. Congrats!

File: src/main/java/com/r307/arbitrader/service/ticker/StreamingTickerStrategy.java [hunk 1]
```diff
@@ -62,21 +66,32 @@ public StreamingTickerStrategy(ErrorCollectorService errorCollectorService) {
...
             .stream()
-            .map(pair -> exchange.getStreamingMarketDataService().getTicker(pair).subscribe(
-                ticker -> {
-                    tickers.computeIfAbsent(exchange, e -> new HashMap<>());
-                    tickers.get(exchange).put(pair, ticker);
+            .map(pair -> {
+                // TODO: Temp fix while https://github.com/knowm/XChange/pull/3761 is not merged and released
+                final Observable<Ticker> tickerObservable;
+                final String exchangeName = exchange.getExchangeSpecification().getExchangeName();
+                if (exchangeName.contains("Gemini")) {
```
Comment (scionaltera) [inline src/main/java/com/r307/arbitrader/service/ticker/StreamingTickerStrategy.java:73]: Could we do something like `if (exchange instanceof GeminiStreamingMarketDataService)` instead, so we don't match the non-streaming Gemini or any other exchanges that might pop up in the future with Gemini in their names? This only affects the Gemini streaming exchange specifically, correct?

Comment (dazito) [reply src/main/java/com/r307/arbitrader/service/ticker/StreamingTickerStrategy.java:73]: That is actually a good idea, I will push it in a minute.
Yes, so far (from all exchanges I tested) it only happened with Gemini Streaming.

**Comment by dazito** [comment]
Is it fine passing in 10 for StreamingMarketDataService#getTicker args parameter? I am not very sure if this is a good value (from Gemini documentation, the 10 will be the limit in the number of elements we process from the order book, as seen from info.bitrich.xchangestream.gemini.GeminiAdaptersX#toOrderbook:

```
  public static OrderBook toOrderbook(
      GeminiOrderbook geminiOrderbook, int maxLevels, Date timestamp) {
    List<LimitOrder> askOrders =
        geminiOrderbook.getAsks().stream()
            .limit(maxLevels) // we will limit to 10 <<<--------------------
            .map(
                (GeminiLimitOrder geminiLimitOrder) ->
                    toLimitOrder(
                        geminiOrderbook.getCurrencyPair(), geminiLimitOrder, Order.OrderType.ASK))
            .collect(Collectors.toList());

    List<LimitOrder> bidOrders =
        geminiOrderbook.getBids().stream()
            .limit(maxLevels) // we will limit to 10 <<<--------------------
            .map(
                (GeminiLimitOrder geminiLimitOrder) ->
                    toLimitOrder(
                        geminiOrderbook.getCurrencyPair(), geminiLimitOrder, Order.OrderType.BID))
            .collect(Collectors.toList());

    return new OrderBook(timestamp, askOrders, bidOrders);
  }
```

**Comment by scionaltera** [comment]
> Is it fine passing in 10 for StreamingMarketDataService#getTicker args parameter? I am not very sure if this is a good value, > the 10 will be the limit in the number of elements we process from the order book

Yeah, 10 seems pretty low. I'm not really sure what a good value would be either, but my gut feeling would be more like 100. It is probably worth testing to see if there is an observable performance impact with higher values.

**Comment by dazito** [comment]
Ok, I have been running the bot for around 48h with the number 10. So far no issues but it also did not perform any trade... I will test with 100.

**Comment by scionaltera** [comment]
The build failure is because the dependency for Gemini is missing in this branch. I haven't "officially" worked on support for the streaming Gemini exchange yet. If it's working well for you this will be a great addition. :)

```
> Task :compileJava
--
107 | /opt/build/src/main/java/com/r307/arbitrader/service/ticker/StreamingTickerStrategy.java:7: error: package info.bitrich.xchangestream.gemini does not exist
108 | import info.bitrich.xchangestream.gemini.GeminiStreamingExchange;
109 | ^
110 | /opt/build/src/main/java/com/r307/arbitrader/service/ticker/StreamingTickerStrategy.java:73: error: cannot find symbol
111 | if (exchange instanceof GeminiStreamingExchange) {
112 | ^
113 | symbol:   class GeminiStreamingExchange
114 | location: class StreamingTickerStrategy
115 | 2 errors
116 |  
117 | > Task :compileJava FAILED
```

**Comment by scionaltera** [comment]
I just checked it looks like XChange 5.0.3 has been released and includes your bugfix:
https://github.com/knowm/XChange/blob/develop/xchange-stream-gemini/src/main/java/info/bitrich/xchangestream/gemini/GeminiStreamingMarketDataService.java#L61

So with that in mind I propose that we simplify this PR down to just the one-liner to fix the missing `exchangeService.convertExchangePair(exchange, pair)` and I'll test and merge in the PRs from dependabot to update the version of XChange. That will avoid the conditional with the TODO on it and bring Arbitrader's version of XChange up to date across the board.

**Comment by dazito** [comment]
The new release of XChange does not have my bug fix :( It still has the bug in there -> `maxDepth` can be null:

`Integer maxDepth = args.length > 0 ? (Integer) args[0] : null;`

Also, the last commit on `GeminiStreamingMarketDataService` is from the 2nd of August: https://github.com/knowm/XChange/blob/master/xchange-stream-gemini/src/main/java/info/bitrich/xchangestream/gemini/GeminiStreamingMarketDataService.java

Maybe my fix is in the new SNAPSHOT version but I don't think we should use the snapshot version.

**Comment by dazito** [comment]
:)

Should I commit the build.gradle with the dependency for gemini streaming or you prefer to wait a few more days?

**Comment by dazito** [comment]
Done, the build should be green now (hopefully)

**Pull request merged**